src/lower/mod.rs

```diff
@@ -1347,21 +1357,36 @@ impl ir::TraitDatum {
 }
 
 impl ir::AssociatedTyDatum {
-    fn to_program_clauses(&self, program: &ir::Program) -> Vec<ir::ProgramClause> {
-        // For each associated type, we define a normalization "fallback" for
-        // projecting when we don't have constraints to say anything interesting
-        // about an associated type.
+    fn to_program_clauses(&self, _program: &ir::Program) -> Vec<ir::ProgramClause> {
+        // For each associated type, we define the "projection
+        // equality" rules. There are always two; one for a successful normalization,
+        // and one for the "fallback" notion of equality.
         //
         // Given:
         //
         //    trait Foo {
         //        type Assoc;
         //    }
         //
-        // we generate:
+        // we generate the 'fallback' rule:
+        //
+        //    forall<T> {
+        //        proj <T as Foo>::Assoc = Foo::Assoc<T>
```

Comment:
I’m not sure why we don’t need a `:- T: Foo` condition here, especially looking at the reverse rule below

Reply:
Good point. Although currently I've hacked out that reverse rule, the comment is out of date, and instead made `T: Foo` part of the "WF" criteria (which I guess is your answer). Still, it seems like otherwise I suppose we would need that.